File: Sliding Window/longest_substring_without_repeating_characters.py
"""
https://leetcode.com/problems/longest-substring-without-repeating-characters/description/
"""
class Solution: #0(n) time complexity, #0(n) space complexity
    def lengthOfLongestSubstring(self, s: str) -> int:
        hashset = set()
        res = 0
        l = 0

        for r in range(len(s)):
            while s[r] in hashset:
                hashset.remove(s[l])
                l += 1
            
            hashset.add(s[r])
            res = max(res, r - l + 1)

        return res

File: Sliding Window/longest_substring_without_repeating_characters.js
"""
https://leetcode.com/problems/longest-substring-without-repeating-characters/description/
"""


class Solution {
    /**
     * @param {string} s
     * @return {number}
     * Time complexity O(n), Space complexity O(n)
     */
    lengthOfLongestSubstring(s) {
        const hashset = new Set();
        var res = 0;
        var l = 0;

        for(let r = 0; r < s.length; r++){
            while(l <= r && hashset.has(s[r])){
                hashset.delete(s[l]);
                l += 1;
            }

            hashset.add(s[r]);
            res = Math.max(res, r - l + 1);
        }

        return res;
    }
}
